### DISCUSSION TITLE
Queued Apex Jobs on RollupAsyncProcessor

### DISCUSSION THREAD

**Discussion by jkenzer** [jamessimone/apex-rollup#139] [Q&A] [answered]
@jamessimone Question for you. Is it normal and expected behavior that there are constantly running Queued and Processing Apex Jobs on the RollUpAsyncProcessor. I ran a full recalculation via CMDT on Friday afternoon and now I'm seeing this behavior non-stop since.

![image](https://user-images.githubusercontent.com/643517/125310566-916b8e80-e2e7-11eb-88bc-296d3d9ebdaf.png)

**Comment by jamessimone** [comment, marked as answer]
That is definitely not expected. Let me take a look at what might be causing that and get back to you.

**Reply by jamessimone** [reply]
@jkenzer I would kill off those recursively queuing jobs. I am going to add in some additional logging and deploy a new version for you so that we can diagnose the problem (is this on a sandbox or production?) - `RollupAsyncProcessor` is only supposed to retry operations up to 100 times, but clearly something is amiss here. In my own testing, I'm not able to reproduce this issue, but with additional logging in place I hope we can nail down where this is occurring.

**Reply by jkenzer** [reply]
@jamessimone Do you know how I go about canceling the recursively queuing jobs? I've aborted jobs, but it doesn't kill the recursively generating part.

**Reply by jamessimone** [reply]
@jkenzer you should be able to use the following anonymous apex script to safely abort the running jobs:

```java
List<AsyncApexJob> queuedJobs = [
    SELECT Id
    FROM AsyncApexJob
    WHERE Status = 'Queued'
    AND ApexClass.Name = 'RollupAsyncProcessor'
    LIMIT 30
];
for (AsyncApexJob job : queuedJobs) {
    try {
        System.abortJob(job.Id);
    } catch (Exception e) { }
}
```

**Reply by jkenzer** [reply]
@jamessimone Thanks for the code. I deleted them all out and then ran a full recalculate in my sandbox again to see if I got the same results. I do get the same behavior. Do you want me to submit this as an issue and if so, what other kind of support can I provide?

**Reply by jamessimone** [reply]
@jkenzer feel free to raise an issue for this - I am also adding in additional logging and potentially some other speculative changes that may solve this; either way, we'll get it sorted out with either the logs or the changes I'm planning to make.

**Reply by jamessimone** [reply]
@jkenzer in the latest release, there are additional logging statements that may help to clarify why rollup is continually re-running for you. Let me know if you want some help poring through them!

**Reply by jkenzer** [reply]
@jamessimone I just encountered this issue again with the latest release trying to do the initial populating on a new rollup. What should I look for in the logs to help determine what's going on?

**Reply by jamessimone** [reply]
@jkenzer I added some logging statements that you should see if you have your `Is Rollup Logging Enabled` checkbox set to true on the Org Default for RollupControl__mdt - things like `Number of deferred rollups (#), still allowed to requeue?` and so on. Hopefully there's enough information in those messages to determine what is causing this endless re-queue scenario.
